CDS/js/Datastore.js

```diff
@@ -0,0 +1,231 @@
+                            reject(err);
+                        } else {
+                            fulfill();
+                        }
+                    });
+                });
+            });
+        },function(err){
+            reject(err);
+        });
+    });
+}
+
+function set(conn, type, key, value) {
+    return setMulti(conn,type,{[key]: value});
+}
+
+function get(conn,type, key, callback, userdata) { /* callback as function(err, value) */
+    if (type !== "meta" && type !== "media" && type !== "tracks") throw "type must be meta, media or track";
+    return new Promise(function (fulfill, reject) {
+        getSource(type, conn.whoami).then(function (sourceid) {
+            db.serialize(function () {
+                var stmt = db.prepare("select value from " + type + " where source_id=? and key=?");
+                stmt.get(sourceid, key, function (err, row) {
+                    if (err) {
+                        console.error(err);
+                        reject(err);
+                    }
+                    if (!row) {
+                        //reject("no row found with key \'" + key + "\'");
+                        fulfill({value: "(value not found)",type: type,key: key});
+                    } else {
+                        //console.log(row);
+                        if(callback){
+                            fulfill(callback(row.value,type,key,userdata));
+                        } else {
+                            fulfill({value: row.value,type: type,key: key});
+                        }
+                    }
+                });
+            });
+        });
+    });
+}
+
+module.exports = {
+    Connection: Connection,
+    newDataStore: function() {
+        return new Promise(function(fulfill, reject) {
+            db.serialize(function () {
+                db.parallelize(function () {
+                    db.run("CREATE TABLE sources (id integer primary key autoincrement,type,provider_method,ctime,filename,filepath)");
+                    db.run("CREATE TABLE meta (id integer primary key autoincrement,source_id,key,value)");
+                    db.run("CREATE TABLE system (schema_version,cds_version)");
+                    db.run("CREATE TABLE tracks (id integer primary key autoincrement,source_id,track_index,key,value)");
+                    db.run("CREATE TABLE media (id integer primary key autoincrement,source_id,key,value)");
+                });
+                db.run("INSERT INTO system (schema_version,cds_version) VALUES (1.0,3.0)", function(err){
+                    if(!err) {
+                        fulfill(this.lastID);
+                    } else {
+                        console.error(err);
+                        reject(err);
+                    }
+                });
+            });
+        });
+    },
+    close: function() {
+        db.close();
+    },
+    set: set,
+    setMulti: setMulti,
+    get: get,
+    substituteString: function(conn,str){
```

Comment:
I've put together an alternative substitute string function moving as much as possible into sync replacements and avoiding unneeded promise creation.

```js
function(conn,str){
        return new Promise(function(fulfill,reject) {
            
            function replaceAllOccurances(string, matchText, replacementValue) {
                //This replaces all occurances of matchText (by splitting) then replaces then with the value (by joining)
                const sanitisedReplacement = replacementValue ? replacementValue : "undefined";
                return string.split(matchText).join(sanitisedReplacement);
            }

            function performStaticSubstitutions(str) {
                const d = new Date();
                const static_subs = {
                    '{route-name}': process.env.cf_route_name,
                    '{hostname}': process.env.HOSTNAME,
                    '{ostype}': process.env.OSTYPE,
                    '{year}': d.getFullYear(),
                    '{month}': d.getMonth()+1,
                    '{day}': d.getDay(),
                    '{hour}': d.getHours(),
                    '{min}': d.getMinutes(),
                    '{sec}': d.getSeconds()
                };

                return Object.keys(static_subs).reduce((currentString, staticSubKey) => {
                    return replaceAllOccurances(currentString, staticSubKey, static_subs[staticSubKey])
                }, str)
            }


            function performConfigSubstitutions(str) {
                return Object.keys(conn.staticDefs).reduce((currentString, currentDef) => {
                    const matchText = "{config:" + currentDef + "}"
                    return replaceAllOccurances(currentString, matchText, conn.staticDefs[currentDef] )
                }, str);
            }

            str = performStaticSubstitutions(str);
            str = performConfigSubstitutions(str);


            /* Perform the Async Datastore Replacements */
            var promiseList = [];
            var param_matcher = /\{(\w+):([^}]+)\}/g;

            while(matches = param_matcher.exec(str)){
                var matchtext = matches[0];
                var type=matches[1];
                var key=matches[2];

                if(type !== 'config') {
                    promiseList.push(get(conn, type, key, function (result, type, key, matchtext) {
                        return {'find': matchtext, 'replace': result}
                    }, matchtext));
                } else {
                    promiseList.push(new Promise((fulfill, reject) => {
                        fulfill({'find': matchtext, 'replace': "(value not found)"});
                    }));
                }
            }

            Promise.all(promiseList).done(function(replacementsList){                
                const returnString = replacementsList.reduce((currentString, replacementObj) => {
                    return replaceAllOccurances(currentString, replacementObj.find, replacementObj.replace);
                }, str)
                fulfill(returnString);
            },function(err){
                reject(err);
            });
        });
    }
```